File: apps/mobile/android/app/src/main/java/com/plainva/app/ShareTargetPlugin.java
package com.plainva.app;

import android.app.Activity;
import android.content.Intent;
import android.widget.Toast;
import com.getcapacitor.*;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@CapacitorPlugin(name = "ShareTarget")
public class ShareTargetPlugin extends Plugin {
    // Intent staging and all bridge mutations are serial: a list cannot mark
    // a currently copying entry interrupted, or race an acknowledgement.
    private static final ExecutorService IO = Executors.newSingleThreadExecutor();
    static void stage(Activity activity, Intent intent, String id) {
        IO.execute(() -> {
            try {
                boolean waiting = new ShareQueue(activity).accept(activity, intent, id);
                if (waiting) activity.runOnUiThread(() -> Toast.makeText(activity, activity.getString(R.string.share_saved), Toast.LENGTH_LONG).show());
            } catch (Exception error) {
                String code = ShareQueue.code(error);
                int message = code.equals("SHARE_LIMIT") ? R.string.share_limit : code.equals("SHARE_QUEUE_FULL") ? R.string.share_full : R.string.share_failed;
                activity.runOnUiThread(() -> Toast.makeText(activity, activity.getString(message), Toast.LENGTH_LONG).show());
            }
            if (activity instanceof MainActivity) {
                MainActivity main = (MainActivity) activity;
                if (main.getBridge() != null) main.getBridge().triggerWindowJSEvent("m-poll-share");
            }
        });
    }
    private interface Work { JSObject run(ShareQueue store) throws Exception; }
    private void execute(PluginCall call, Work work) {
        IO.execute(() -> {
            try { call.resolve(work.run(new ShareQueue(getContext()))); }
            catch (Exception error) { call.reject(ShareQueue.code(error)); }
        });
    }
    @PluginMethod public void listPendingShares(PluginCall call) {
        execute(call, store -> { JSObject result = new JSObject(); result.put("entries", store.list()); return result; });
    }
    @PluginMethod public void readFileChunk(PluginCall call) {
        execute(call, store -> {
            Integer offset = call.getInt("offset"), length = call.getInt("length");
            if (offset == null || length == null) throw new Exception("SHARE_INVALID");
            JSObject result = new JSObject(); result.put("data", store.chunk(call.getString("id"), call.getString("fileId"), offset, length)); return result;
        });
    }
    @PluginMethod public void beginImport(PluginCall call) {
        execute(call, store -> {
            JSObject plan = call.getObject("plan"); if (plan == null) throw new Exception("SHARE_INVALID");
            JSObject result = new JSObject(); result.put("entry", store.begin(call.getString("id"), plan)); return result;
        });
    }
    @PluginMethod public void markImported(PluginCall call) {
        execute(call, store -> { store.mark(call.getString("id"), call.getString("fileId"), Boolean.TRUE.equals(call.getBoolean("note"))); return new JSObject(); });
    }
    @PluginMethod public void finishShare(PluginCall call) {
        execute(call, store -> { store.finish(call.getString("id"), Boolean.TRUE.equals(call.getBoolean("discard"))); return new JSObject(); });
    }
}

File: apps/mobile/ios/App/App/ShareTargetPlugin.swift
import Foundation
import Capacitor

@objc(ShareTargetPlugin)
public class ShareTargetPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ShareTargetPlugin"
    public let jsName = "ShareTarget"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "listPendingShares", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "readFileChunk", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "beginImport", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "markImported", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "finishShare", returnType: CAPPluginReturnPromise)
    ]
    private let io = DispatchQueue(label: "com.plainva.share-inbox")
    private func run(_ call: CAPPluginCall, action: @escaping (ShareQueueStore) throws -> [String: Any]) {
        io.async {
            do { call.resolve(try action(ShareQueueStore())) }
            catch { call.reject((error as? ShareQueueFailure)?.rawValue ?? "SHARE_STORAGE") }
        }
    }
    @objc func listPendingShares(_ call: CAPPluginCall) {
        run(call) { store in ["entries": try store.list()] }
    }
    @objc func readFileChunk(_ call: CAPPluginCall) {
        run(call) { store in
            guard let id = call.getString("id"), let fileId = call.getString("fileId"), let offset = call.getInt("offset"), let length = call.getInt("length") else { throw ShareQueueFailure.invalid }
            return ["data": try store.chunk(id, fileId: fileId, offset: offset, length: length)]
        }
    }
    @objc func beginImport(_ call: CAPPluginCall) {
        run(call) { store in
            guard let id = call.getString("id"), let plan = call.getObject("plan") else { throw ShareQueueFailure.invalid }
            return ["entry": try store.beginImport(id, plan: plan)]
        }
    }
    @objc func markImported(_ call: CAPPluginCall) {
        run(call) { store in
            guard let id = call.getString("id") else { throw ShareQueueFailure.invalid }
            try store.mark(id, fileId: call.getString("fileId"), note: call.getBool("note") ?? false)
            return [:]
        }
    }
    @objc func finishShare(_ call: CAPPluginCall) {
        run(call) { store in
            guard let id = call.getString("id") else { throw ShareQueueFailure.invalid }
            try store.finish(id, discard: call.getBool("discard") ?? false)
            return [:]
        }
    }
}
